pymc_extras/statespace/models/DFM.py

```diff
@@ -0,0 +1,828 @@
+
+            # Observation noise covariance matrix
+            obs_sigma = pm.HalfNormal("sigma_obs", dims=["k_endog"])
+
+            # Build the symbolic graph and attach it to the model
+            dfm_mod.build_statespace_graph(data=data, mode="JAX")
+
+            # Sampling
+            idata = pm.sample(
+                draws=500,
+                chains=2,
+                nuts_sampler="nutpie",
+                nuts_sampler_kwargs={"backend": "jax", "gradient_backend": "jax"},
+            )
+
+    """
+
+    def __init__(
+        self,
+        k_factors: int,
+        factor_order: int,
+        k_endog: int | None = None,
+        endog_names: Sequence[str] | None = None,
+        k_exog: int | None = None,
+        exog_names: Sequence[str] | None = None,
+        shared_exog_states: bool | None = None,
+        exog_innovations: bool | None = None,
+        error_order: int = 0,
+        error_var: bool = False,
+        error_cov_type: str = "diagonal",
+        measurement_error: bool = False,
+        verbose: bool = True,
+    ):
+        if k_endog is None and endog_names is None:
+            raise ValueError("Either k_endog or endog_names must be provided.")
+        if k_endog is None:
+            k_endog = len(endog_names)
+        if endog_names is None:
+            endog_names = [f"endog_{i}" for i in range(k_endog)]
+
+        self.endog_names = endog_names
+        self.k_endog = k_endog
+        self.k_factors = k_factors
+        self.factor_order = factor_order
+        self.error_order = error_order
+        self.error_var = error_var
+        self.error_cov_type = error_cov_type
+
+        if k_exog is None and exog_names is None:
+            self._exog = False  # flag for presence of exogeneous variables
+            self.k_exog = 0
+        else:
+            self._exog = True
+            self.shared_exog_states = shared_exog_states
+            self.exog_innovations = (
+                exog_innovations if exog_innovations is not None else False
+            )  # default if not provided is False
+            if k_exog is None:
+                k_exog = len(exog_names) if exog_names is not None else 0
+            elif exog_names is None:
+                exog_names = [f"exog_{i}" for i in range(k_exog)] if k_exog > 0 else None
+            self.k_exog_states = k_exog * k_endog if not shared_exog_states else k_exog
+
+        self.exog_names = exog_names
+        self.k_exog = k_exog
+
+        # TODO add exogenous variables support
+        # I start implementing a version of exog support based on pymc_extras/statespace/models/structural/components/regression.py
+        # exog_innovations control if the beta coefficient follows a random walk
+        # shared_exog_states control if the exogenous states are shared across equations
+        # I tested the case of shared_exog_states=True and exog_innovations=False vs the stats case by looking at trajectory and everything works well
```

Comment:
Excellent! But no need to keep the comments. Make a unittest to reproduce your test.

Keep a comment noting the difference between the statsmodels implementation (in the obs vector) and ours (in the state vector)